src/dmd/backend/dvec.d

```diff
@@ -0,0 +1,605 @@
+ *
+ * Copyright:   Copyright (c) 2013-2018 by The D Language Foundation, All Rights Reserved
+ * Authors:     $(LINK2 http://www.digitalmars.com, Walter Bright)
+ * License:     $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Source:      $(LINK2 https://github.com/dlang/dmd/blob/master/src/dmd/backend/dvec.d, backend/dvec.d)
+ */
+
+module dvec;
+
+import core.stdc.stdio;
+import core.stdc.stdlib;
+import core.stdc.string;
+
+import core.bitop;
+
+extern (C):
+
+nothrow:
+@nogc:
+
+alias size_t vec_base_t;                     // base type of vector
+alias vec_base_t* vec_t;
+
+enum VECBITS = vec_base_t.sizeof * 8;        // # of bits per entry
+enum VECMASK = VECBITS - 1;                  // mask for bit position
+enum VECSHIFT = (VECBITS == 16) ? 4 : (VECBITS == 32 ? 5 : 6);   // # of bits in VECMASK
+
+static assert(vec_base_t.sizeof == 2 && VECSHIFT == 4 ||
+              vec_base_t.sizeof == 4 && VECSHIFT == 5 ||
+              vec_base_t.sizeof == 8 && VECSHIFT == 6);
+
+struct VecGlobal
+{
+    int count;           // # of vectors allocated
+    int initcount;       // # of times package is initialized
+    vec_t[30] freelist;  // free lists indexed by dim
+
+  nothrow:
+  @nogc:
+
+    void initialize()
+    {
+        if (initcount++ == 0)
+            count = 0;
+    }
+
+    void terminate()
+    {
+        if (--initcount == 0)
+        {
+            debug
+            {
+                if (count != 0)
+                {
+                    printf("vecGlobal.count = %d\n", count);
+                    assert(0);
+                }
+            }
+            else
+                assert(count == 0);
+
+            foreach (size_t i; 0 .. freelist.length)
+            {
+                void **vn;
+                for (void** v = cast(void **)freelist[i]; v; v = vn)
+                {
+                    vn = cast(void **)(*v);
+                    //mem_free(v);
+                    .free(v);
+                }
+                freelist[i] = null;
+            }
+        }
+    }
+
+    vec_t allocate(size_t numbits)
+    {
+        if (numbits == 0)
+            return cast(vec_t) null;
+        const dim = (numbits + (VECBITS - 1)) >> VECSHIFT;
+        vec_t v;
+        if (dim < freelist.length && (v = freelist[dim]) != null)
+        {
+            freelist[dim] = *cast(vec_t *)v;
+            v += 2;
+            switch (dim)
+            {
+                case 5:     v[4] = 0;  goto case 4;
+                case 4:     v[3] = 0;  goto case 3;
+                case 3:     v[2] = 0;  goto case 2;
+                case 2:     v[1] = 0;  goto case 1;
+                case 1:     v[0] = 0;
+                            break;
+                default:    memset(v,0,dim * vec_base_t.sizeof);
+                            break;
+            }
+            goto L1;
+        }
+        else
+        {
+            v = cast(vec_t) calloc(dim + 2, vec_base_t.sizeof);
+            assert(v);
```

Comment:
This should better be reporting "out of memory", the backend uses `err_nomem()` for this.

Reply:
It uses the assert because the file is supposed to be independent of any other project modules. The C++ version it was translated from did this, too.

Reply:
The C++ version uses mem_calloc/mem_exception from mem.c.

`assert(v)` has the problem that it is likely to go away in a release build, so if you want no dependencies I'd recommend `if (!v) assert(false, "out of memory")`.

Reply:
See my other comment about why assert not being disabled is required in the back end.